podium-gateway/src/main/webapp/app/shared/role/role-assign/role-assign.component.ts

```diff
@@ -0,0 +1,329 @@
+            orgUser.searchTerm = orgUser.fullName;
+        }
+
+        if ( role ) {
+            orgUser.previousAuthority = role.authority;
+            orgUser.authority = role.authority;
+            orgUser.isSaved = user.uuid != null;
+        }
+
+        orgUser.dataSource = this.getDatasourceForUser(orgUser);
+
+        return orgUser;
+    }
+
+    /**
+     * Initialise a datasource for an organisation user.
+     * Used for the user lookup typeahead
+     * @param organisationUser
+     * @returns {Observable<R>}
+     */
+    getDatasourceForUser(organisationUser?: OrganisationUser): Observable<any> {
+        /**
+         * User typeahead datasource observable
+         */
+        return Observable.create((observer: any) => {
+            // Runs on every search
+            // each organisation user has a 'searchTerm' property that is used as the input and is bound to [(ngModel)]
+            // when user types into the input .next is called with the value from the input
+            observer.next({query: organisationUser.searchTerm });
+        }).mergeMap((term: any) => this.userService.suggest(term));
+    }
+
+    public save(user: OrganisationUser) {
+        // Find and Update role by authority
+        let role = this.getRoleByAuthority(user.authority);
+        this.updateRole(role, user, false).subscribe(
+            (res) => { this.onSaveSuccess(res, false); },
+            (err) => { this.onError(err); }
+        );
+    }
+
+    public update(user: OrganisationUser) {
+        // Remove previous role
+        let previousRole = this.getRoleByAuthority(user.previousAuthority);
+
+        this.updateRole(previousRole, user, true).subscribe(
+            (previousRes: Response) => {
+                // Add new role
+                let role = this.getRoleByAuthority(user.authority);
+                this.updateRole(role, user, false).subscribe(
+                    (res: Response) => { this.onSaveSuccess(res, false); },
+                    (err: Response) => { this.onError(err); },
+                );
+            },
+            (err: Response) => { this.onError(err); }
+        );
+    }
+
+    public delete(user: OrganisationUser) {
+        // Find and update role by authority
+        let role = this.getRoleByAuthority(user.authority);
+
+        this.updateRole(role, user, true).subscribe(
+            (res: Response) => { this.onSaveSuccess(res, true); },
+            (err: Response) => { this.onError(err); }
+        );
+    }
+
+    public loadAllRolesForOrganisation(): Observable<any> {
+        return Observable.create((observer: any) => {
+
+            this.organisationUsers = [];
+            if (this.organisation) {
+                this.roleService.findAllRolesForOrganisation(this.organisation.uuid).subscribe(roles => {
+                    this.organisationRoles = roles;
+                    for (let i = 0; i < roles.length; i++) {
+                        let role: Role = roles[i];
```

Comment:
too many level of nested code and on each level has async calls. Consider to break this chunk into smaller functions so that it's easier to understand